### ISSUE TITLE
inconsistency meaning

### ISSUE THREAD

**Issue by btwewek** [shap/shap#647]
Hi,

I am trying to understand consistent problem of feature importance.
If I using the same dataset, and fitting two different parameters of model(e.g., xgboost),
then I got two different mean|SHAP| values where the ranking of feature were diffferent.
Does this mean inconsistency?

1.
```

X_train,X_test,Y_train,Y_test = train_test_split(*shap.datasets.iris(), test_size=0.2, random_state=0)
model = xgboost.XGBClassifier(n_estimators=50, max_depth=12)
model.fit(X_train, Y_train)
```
![image](https://user-images.githubusercontent.com/46321291/59558512-8d94b180-9026-11e9-83e1-29de292972db.png)

2.
```
model = xgboost.XGBClassifier(n_estimators=100, max_depth=3)
model.fit(X_train, Y_train)
```
![image](https://user-images.githubusercontent.com/46321291/59558517-a735f900-9026-11e9-83c5-795de4f95458.png)

**Comment by slundberg** [comment]
A model (such as XGBoost) is just a way to approximate an unkown a functional relationship found in a dataset. Different training parameters will lead to different models and those models may depend a bit differently on different input features. This is what you are seeing. "Inconsistency" as used when defining the Shapley values is a case where the importance measure does not match what the model depends on. The differences you are seeing in this example are not this, but rather are due to real differences between the models (which I can say because TreeExplainer has consistency guarantees).

**Comment by btwewek** [comment]
Thanks for the answer, it is helpful.
So in [Tree SHAP arXiv paper](https://arxiv.org/abs/1802.03888) Fig.1(below), there are two models were trained on the **same dataset**, the output` [Cough&Fever]*80` and` [Cough&Feveer]*80+Cough*10 `are model's prediction (not the ground truth of dataset cause we don't know real ground truth).
![image](https://user-images.githubusercontent.com/46321291/59679183-be304300-9200-11e9-8e23-88681beb12d1.png)

Inconsistency of Gain and Split count in the Model_B is because feature importance ranking is switched (We expected importance of Cough should be greater than Fever). 
I thought the Gain and Split count are inconsistency due to the different feature importance ranking between Model_A and Model_B.
I am not sure if there is any misunderstanding.

**Comment by slundberg** [comment]
In that figure we are showing an example where the model is so simple that we can tell that Model B always depends more on Cough than Model A does. Model B depends more on Cough because we trained it on a dataset with different labels (ones that cause the model to depend more on cough).

While being inconsistent can (and often does) cause the importance ranking to be wrong, it is fundamentally a question of if we can find two models where the second model depends more on a specific feature than the first model (in a very obvious sense made precise by the axiom), but that feature gets less credit in the second model than the first. This is what we see in that figure.

**Comment by btwewek** [comment]
> Model B depends more on Cough because we trained it on a dataset with different labels 

So actually Model A and B are trained on different target, if that, why could we compare feature importance between these two models?

Can I just judge from output of Model B : `[Cough&Fever]*80+Cough*10` which is obviously that `Cough` important than `Fever` but result of Gain is not, so Gain is inconsistent?

Sorry about this issue doesn't helpful for SHAP package growth, but I'm really want to figure out.

**Comment by slundberg** [comment]
Unless you have a random training process you will always get the same model when you train on the same labels, and comparing models is really only interesting when they are different. In this case I think the best way to think about it is as you said, Model B clearly has a higher dependence on Cough than Model A (just by intuitive inspection), yet the gain credit is lower for Cough in Model B. There is a way to make that intuitive inspection mathematically precise, but it is just the same idea and much more involved.

**Comment by varunmahajan53** [comment]
@slundberg ,

**Comment by varunmahajan53** [comment]
@slundberg , 
I have a similar kind of issue. 
I am getting the inconsistent feature's contribution for same input dataset. Contributions are always different in every run. We are working on the compliance project and consistency of  contributions are very crucial for us.

explainer = shap.SamplingExplainer(fn_predict, data=val_data_sampled_enc)
shap_values = explainer.shap_values(x_test_enc, silent=True)

```
run 1:

[[-1.92340864e-03 -1.45998722e-02 -6.18119664e-03  1.02872582e-02
  -1.74573145e-02 -1.45659983e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -3.62246202e-02  1.83270444e-05 -1.55354485e-04
   7.49064341e-02  1.64473664e-03]
 [-2.95104268e-03  6.57229807e-03  6.42444524e-03 -1.81398802e-03
   2.86218328e-02  6.62192215e-03  0.00000000e+00  0.00000000e+00
   0.00000000e+00  2.75990792e-02  2.99686422e-03 -2.41660230e-03
  -6.97848579e-02 -1.33932686e-05]
 [ 2.18968971e-03  8.33590736e-05  4.00694505e-05  3.41121673e-05
   2.28105971e-04 -2.60735913e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -1.71357393e-02 -2.51246100e-03 -1.21762331e-05
   3.86802728e-02  0.00000000e+00]
 [ 3.58755487e-03 -4.03059747e-03  8.01946721e-05  1.00337345e-04
  -2.88088425e-02 -9.05932387e-03  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -4.99133028e-02  2.48247180e-03  3.44383427e-03
   8.03027593e-02 -2.71625106e-06]
 [-2.95729646e-03  7.33775734e-03  1.12694615e-04 -2.15299983e-03
   1.46632950e-02  3.58479094e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00  8.22503734e-02 -1.12112638e-03 -2.73289420e-03
  -1.19893206e-01 -2.66406907e-03]]
  
  run 2:
  __________
  
  [[-3.12780203e-03 -1.29945942e-02 -4.14822862e-03  4.58461309e-03
  -1.17602317e-02 -1.01264606e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -3.61236927e-02 -2.40549330e-04  1.81183932e-03
   6.61436102e-02  1.73048694e-03]
 [-4.43730736e-03  1.75987414e-03  1.14353230e-03 -1.25848149e-03
   3.25531727e-02  6.10445763e-03  0.00000000e+00  0.00000000e+00
   0.00000000e+00  4.11956831e-02  1.07386240e-03 -1.71306547e-03
  -7.43305158e-02 -2.34649483e-04]
 [ 3.46494409e-03  1.07803908e-03  4.71439554e-05  8.61547603e-05
  -3.11543850e-03 -2.18266660e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -2.02742201e-02 -1.13571622e-03 -7.58796564e-05
   3.72732787e-02  0.00000000e+00]
 [ 5.97316913e-04 -3.44543074e-03  6.03716738e-05  2.19530342e-04
  -1.68288635e-02 -1.04313779e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00 -6.02822552e-02  1.14449244e-03  4.42591132e-03
   8.27284858e-02 -5.81105228e-06]
 [-8.44876222e-04  2.48200040e-03  1.11672329e-03 -6.32371321e-03
   1.88703786e-02  4.14449528e-02  0.00000000e+00  0.00000000e+00
   0.00000000e+00  7.28325181e-02 -2.13781082e-03 -2.36008270e-04
  -1.18338586e-01 -1.75141247e-04]]
  ```
  Is there any way where we can control the randomness of data and get always consistent contributions everytime we run the explainer ??
